### PR TITLE
Load Average for KVM

### PR DESCRIPTION
**Pull request by DennisKonrad**
## Description
This adds the **load average for the kvm hypervisor**. The average Load already seems to be available for Xen.

In the _listHosts_ the [averageload](https://cloudstack.apache.org/api/apidocs-4.13/apis/listHosts.html) is already present. In _listHostMetrics_ it appears as "cpuloadaverage".

It adds the 1-min load average from /proc/loadavg to the
_listHosts
listHostsMetrics_
API calls.

image

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
Built and tested in kvm cluster.
Tested manually via API calls

### CODE REVIEW

File: plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java [hunk 1]
```diff
@@ -43,6 +43,10 @@
     @Param(description = "the total cpu allocated in Ghz")
     private String cpuAllocated;
 
+    @SerializedName("loadAverage")
+    @Param(description = "the average cpu load the last minute")
+    private String loadAverage;
```
Comment (wido) [inline plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java]: Shouldn't this be a Double?

Comment (DennisKonrad) [reply plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java]: I don't think so. I think it's just for output. The other variables here are either Booleans or Strings

Comment (wido) [reply plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java]: Ah, yes. I misread. This is the API response to the outside world.

**Comment by GabrielBrascher** [comment]
@blueoragutan package

**Comment by GabrielBrascher** [comment]
@blueorangutan package

File: plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java [hunk 2]
```diff
@@ -43,6 +43,10 @@
     @Param(description = "the total cpu allocated in Ghz")
     private String cpuAllocated;
 
+    @SerializedName("loadAverage")
```
Comment (weizhouapache) [inline plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java]: is cpuloadaverage better ?

**Comment by weizhouapache** [comment]
@DennisKonrad 
I have tested it , the value it looks fine.
The only problem is, I spent some time to look for the value in the api responses.  it would be better to change them to same name

(local) 🐵 > list hosts filter=averageload
(local) 🐵 > list hostsmetrics filter=loadAverage

File: plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java (hunk 2 above)
Comment (DennisKonrad) [reply plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java]: you are right, will change it to that

Comment (DennisKonrad) [reply plugins/metrics/src/main/java/org/apache/cloudstack/response/HostMetricsResponse.java]: For easier usage I changed both the results in 
list hosts
list hostsmetrics
to cpuloadaverage.

**Comment by DennisKonrad** [comment]
@weizhouapache I changed both of the keys to cpuloadaverage:

image

image

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
I would start a regression suite, but...
Do we need test automation for this?
/cc @weizhouapache @GabrielBrascher @DennisKonrad

**Comment by GabrielBrascher** [comment]
@DaanHoogland I don't think that automation tests are strictly necessary in this case. It is a simple change, and we already have code reviews and one manual tests by @weizhouapache.

I am assuming that the centos6 failing to package is not related to this PR.

**Comment by wido** [comment]
@DennisKonrad Why do the Host Metrics on your screenshots show a String and not a Double?

We should return this in the API as a Double and not a String.

**Comment by GabrielBrascher** [comment]
Good point @wido. It is set as double or Double on `HostStats.java`, `HostResponse.java`, `GetHostStatsAnswer.java`, and `HostStatsEntry.java`. Additionally, at `CPUStat.java` the method `getCpuLoadAverage()` returns Double.

Therefore, it looks OK on the PR. However, `list hostmetrics` returns it as String; `list hosts` looks fine though. I will take a look around the `list hostmetrics`.

On a 5.0 we can normalize API values and fix issues such as % values returned as Strings instead of double.

**Comment by DennisKonrad** [comment]
@wido Good catch. Will change that to output as double

**Comment by DaanHoogland** [comment]
@GabrielBrascher centos6 is broken on a lot (most if not all) PRs. not to worry about that one.
so a package after @DennisKonrad 's next addition will be enough to carry on.

**Comment by DennisKonrad** [comment]
@wido I think it's better now.

image

**Comment by wido** [comment]
Looking good!

File: api/src/main/java/org/apache/cloudstack/api/response/HostResponse.java [hunk 3]
```diff
@@ -114,9 +114,9 @@
     @Param(description = "the amount of the host's CPU after applying the cpu.overprovisioning.factor ")
     private String cpuWithOverprovisioning;
 
-    @SerializedName("averageload")
+    @SerializedName("cpuloadaverage")
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/response/HostResponse.java]: Can you move this to APIConstants, also changes in the API key response breaks the response compatibility if xen-specific users are using the key for something in their scripts, tool, external integrations. Can you revert the change?

Comment (DennisKonrad) [reply api/src/main/java/org/apache/cloudstack/api/response/HostResponse.java]: It seems to me the guys with xen can't use it anyway:
https://github.com/apache/cloudstack/blob/1d05fead49f5c856257a741b07122f5633d2e359/plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java#L2258-L2272

I will add it to the APIConstanst and leave the API key changed for consistency. Sounds good to you @rhtyd

Comment (rhtyd) [reply api/src/main/java/org/apache/cloudstack/api/response/HostResponse.java]: I'm okay with your rename change if the field is not set previously.

Comment (DennisKonrad) [reply api/src/main/java/org/apache/cloudstack/api/response/HostResponse.java]: Okay, I think [this](https://github.com/apache/cloudstack/pull/3738/commits/e05667692505989001764324a165831420628b05) should do it. I couldn't find any code where the old name is used

**Comment by GabrielBrascher** [comment]
Looking good to merge.
Merging it considering that all checks have passed, we have 4 LGTMs, a couple of manual tests, and the packaging results are ok for Ubuntu and CentOS.
